klippy/extras/temperature_probe.py

```diff
@@ -0,0 +1,375 @@
+        )
+        self.resting_z = config.getfloat("resting_z", .4, above=0.)
+        self.cal_pos = config.getfloatlist(
+            "calibration_position", None, count=3
+        )
+        self.cal_bed_temp = config.getfloat(
+            "calibration_bed_temp", None, above=50.
+        )
+        self.cal_extruder_temp = config.getfloat(
+            "calibration_extruder_temp", None, above=50.
+        )
+        self.cal_extruder_z = config.getfloat(
+            "extruder_heating_z", 50., above=0.
+        )
+        # Setup temperature sensor
+        smooth_time = config.getfloat("smooth_time", 2., above=0.)
+        self.inv_smooth_time = 1. / smooth_time
+        self.min_temp = config.getfloat(
+            "min_temp", KELVIN_TO_CELSIUS, minval=KELVIN_TO_CELSIUS
+        )
+        self.max_temp = config.getfloat(
+            "max_temp", 99999999.9, above=self.min_temp
+        )
+        pheaters = self.printer.load_object(config, "heaters")
+        self.sensor = pheaters.setup_sensor(config)
+        self.sensor.setup_minmax(self.min_temp, self.max_temp)
+        self.sensor.setup_callback(self._temp_callback)
+        pheaters.register_sensor(config, self)
+        self.temp_lock = Lock()
+        self.smoothed_temp = 0.
+        self.last_temp_read_time = 0.
+        self.measured_min = 99999999.
+        self.measured_max = 0.
+
+        # Calibration State
+        self.cal_helper = None
+        self.next_auto_temp = 99999999.
+        self.target_temp = 0
+        self.expected_count = 0
+        self.sample_count = 0
+        self.in_calibration = False
+        self.step = 2.
+        self.last_zero_pos = None
+        self.total_expansion = 0
+        self.start_pos = []
+
+        # Register GCode Commands
+        pname = self.name.split(maxsplit=1)[-1]
+        self.gcode.register_mux_command(
+            "PROBE_DRIFT_CALIBRATE", "PROBE", pname,
+            self.cmd_PROBE_DRIFT_CALIBRATE,
+            desc=self.cmd_PROBE_DRIFT_CALIBRATE_help
+        )
+
+    def register_calibration_helper(self, helper):
+        self.cal_helper = helper
+
+    def _temp_callback(self, read_time, temp):
+        if not temp:
+            return
+        with self.temp_lock:
+            time_diff = read_time - self.last_temp_read_time
+            self.last_temp_read_time = read_time
+            temp_diff = temp - self.smoothed_temp
+            adj_time = min(time_diff * self.inv_smooth_time, 1.)
+            self.smoothed_temp += temp_diff * adj_time
+            self.measured_min = min(self.measured_min, self.smoothed_temp)
+            self.measured_max = max(self.measured_max, self.smoothed_temp)
+            st = self.smoothed_temp
+            if self.in_calibration and st >= self.next_auto_temp:
+                self.next_auto_temp = 99999999.
+                self.printer.get_reactor().register_async_callback(
+                    lambda e, gc=self.gcode: gc.run_script("PROBE_DRIFT_NEXT")
+                )
```

Comment:
This looks like it could have subtle race conditions (as `_temp_callback()` runs in a background thread).

The `self.temp_lock` is only ever taken here, which doesn't seem correct.  (The lock only works if other places in the code also acquire the lock.)

Many of the variables (measured_min, measured_max, in_calibration, next_auto_temp) could change mid-execution (both here and in `get_status()` etc) due to the background thread.

Also, technically, it's possible for PROBE_DRIFT_NEXT to be queued from `_temp_callback()` at a similar moment that the user manually calls `PROBE_DRIFT_NEXT` - which could cause two back-to-back executions.

Reply:
Ah, I must have had async programming on my mind.  FWIW, its probably not a great plan to acquire a thread lock on the main thread since acquisition can be a blocking operation.  I wonder if it would be a good idea to create a wrapper around the lock using the reactor, testing if the lock is held and pausing until it isn't.

Reply:
I'm not sure the reactor can help much as it only has limited thread safety itself (only `register_async_callback()` can be called from the background thread) and the background thread isn't running a reactor (only the main thread has it).

One possible alternative is to store groups of values in tuples, as Python then assures each assignment is atomic.  The clocksync.py code does this.

Something like (totally untested):
```python
    def _temp_callback(self, read_time, temp):
        time_diff = read_time - self.last_temp_read_time
        self.last_temp_read_time = read_time
        smoothed_temp, measured_min, measured_max = self.last_measurement
        temp_diff = temp - smoothed_temp
        adj_time = min(time_diff * self.inv_smooth_time, 1.)
        smoothed_temp += temp_diff * adj_time
        measured_min = min(measured_min, smoothed_temp)
        measured_max = max(measured_max, smoothed_temp)
        self.last_measurement = (smoothed_temp, measured_min, measured_max)
        if self.in_calibration and smoothed_temp >= self.next_auto_temp:
            self.printer.get_reactor().register_async_callback(
                self._check_kick_next)

    def _check_kick_next(self, eventtime):
        smoothed_temp, measured_min, measured_max = self.last_measurement
        if self.in_calibration and smoothed_temp >= self.next_auto_temp:
            self.next_auto_temp = 99999999.
            self.gcode.run_script("TEMPERATURE_PROVE_NEXT")
```

The idea here is that smoothed_temp, measured_min, measured_max are always atomically written as a group, and thus any readers will get those values as a group (never a mix from two possible `_temp_callback()` invocations).  Also, a "dirty read" of self.in_calibration and self.next_auto_temp isn't an issue as long as the values are rechecked in the main thread.